site/docs/design-docs/012-improved-error-handling.md

```diff
@@ -0,0 +1,113 @@
+submitted it and give enough information to facilitate the repair of the
+underlying root cause.
+
+For this reason we should introduce a new `JobStatus`, called
+`SUSPENDED_DUE_TO_ERROR` and provide details in the `Exception` set in
+the `CompletableFuture` returned by `job.getFuture()` (also used by
+`job.join()`).
+
+Since we will use a new state we need to make sure that it behaves
+exactly like the regular suspend: the job can be cancelled, snapshots
+can be exported (enterprise feature) and so on.
+
+## When to use
+
+Do we always want to suspend jobs when there is a failure? It doesn't
+make sense to do it for failures that can't be remedied, but which are
+those? Hard to tell, hard to exhaust all the possibilities.
+
+Since the suspend feature will be optional and will need explicit
+setting, it's probably ok to do it for any failure (not just some
+whitelisted ones). There is nothing to loose anyways. If the failure
+can't be remedied, then the user can just cancel the job. They will be
+no worse off than if it had failed automatically.
+
+## Processing guarantees
+
+Should we enable this feature only for jobs with processing guarantees?
+The answer is probably yes. Jobs without a processing guarantee don't
+save snapshots, so there's not much state to loose. But is this true? Do
+jobs have state outside snapshots?
+
+Plus there is not much to gain by not allowing this feature for
+"stateless" jobs... Maybe we should not discriminate on these grounds.
+
+## Batch jobs
+
+Batch jobs can always just be restarted, as the mantra goes. But still,
+they can take a long time to restart, so we probably should allow them
+to work in this way too.
+
+As with other aspects mentioned above we don't gain anything by
+disabling this functionality for batch jobs.
+
+## Job upgrade
+
+Even if we enable this feature some fixable problems won't be actually
+possible to handle in open-source Jet, because they might require
+pipeline code change and that will need snapshot export and job upgrades
+which are enterprise features.
+
+Not sure what we can do about this, besides being aware of it.
+
+## Failure snapshot
+
+Ideally, when an error happens, which will be handled by suspending the
+job, we would prefer to make all processors take a snapshot right then
+so that we can later resume execution from the most advanced state. But
+this, unfortunately doesn't seem possible.
+
+Snapshots can be taken only when all processors are functioning properly
+(due to the nature of how distributed snapshots happen).
+
+But, even some slightly obsolete snapshot should be better than loosing
+the whole computation, so I guess this is a weakness we can live with.
+
+## In-processor error handling
+
+This functionality should be a solution of last resort, meaning that all
+errors that can be handled without user intervention should be handled
+automatically. For example sources and sinks loosing connection to
```

Comment:
Here we can distinguish streaming and batch sources. Streaming sources are typically run in a fault-tolerant job, batch sources are not. So streaming sources should attempt to reconnect as much as possible or throw `RestartableException`. Batch sources, on the other hand, should not attempt to retry. When you process a remote file and that file is not accessible, there's no point in retrying. Users would be surprised to submit such a job, see it running, and then, after an hour, figure out that the job just keeps trying to connect.

Reply:
I don't agree with _Batch sources, on the other hand, should not attempt to retry._.

If you have a batch job which takes several days to finish, you don't want the job to fail in the middle because of some intermittent network hiccup, which can be easily handled by a retry couple of seconds later.

Reply:
Yeah, I agree. But for a streaming source it's a must. For a batch source it's nice to have.

Reply:
Source behaviour will be independent of this feature. I think we can and should make their behaviour configurable. They should/will be able to attempt retries to a certain degree, but at the end, their solution of last resort will be throwing an exception.